Automattic/jetpack
# Add `jetpack draft` command

This comes out of a discussion that @kraftbj and I had in Slack about how to do the right amount of checks at the right time (linting, tests, changelogger, etc). 

The reality is that sometimes you're doing a lot of collaborative refactoring or working on a draft PR and you want to share your changes without pleasing all the scripts, but we still want to make sure the more serious checks run once you need them.

We also don't want to encourage developers to use the `--no-verify` flag, since some checks may be super important and we don't want them to be skipped.

For this situation, there is now Jetpack Draft Mode. The goal is to enable developers to reach maximum velocity for the type of goal they're trying to accomplish.

To enable:

```
$ jetpack draft enable
```

To disable:

```
$ jetpack draft disable
```

Currently, this has the following effects on pre-commit:

* raise the number of allowable warnings in eslint and eslint-changed to 100
* don't bail if phpcs fails
* don't bail if phplinter fails
* don't bail if renovate changes are missing

And the following effects on pre-push:

* don't bail if changelog files are missing

You will still see all the output from these various commands, it's just that fewer of them will block the commit or push from succeeding.

In addition, when you run `jetpack draft disable`, it will offer to run the pre-commit checks for you right away so you can start fixing errors. It does not run the pre-push checks because for some reason those were hanging.

#### Changes proposed in this Pull Request:

* Add `jetpack draft enable|disable` command for temporarily lightening some pre-commit and pre-push checks.

#### Jetpack product discussion
Not available yet. Perhaps in the comments on this PR?

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
First, run `jetpack cli link` to make sure everything's working.

* Run `jetpack draft enable`. 
* Check that a file .jetpack-draft exists in your repo root
* Run `jetpack draft disable`. 
* Check .jetpack-draft is gone from your repo root

The following test requires that you first fork this branch with `git checkout -b my-jetpack-draft-test` - otherwise you will create accidental commits!

With draft mode enabled, try committing some code that doesn't pass eslint (e.g. a function without a docblock). It shouldn't stop you. Try doing the same with draft mode disabled. You should be blocked from committing.
